IOPool/Input/src/RootTreeCacheManager.cc

```diff
@@ -0,0 +1,544 @@
+    void SparseReadCache::createPrimaryCache(unsigned int cacheSize) {
+      cacheSize_ = cacheSize;
+      treeCache_ = createCacheWithSize(cacheSize);
+      if (treeCache_)
+        treeCache_->SetEnablePrefetching(enablePrefetching_);
+      rawTreeCache_.reset();
+    }
+
+    TTreeCache* SparseReadCache::checkTriggerCache(TBranch* branch, EntryNumber entryNumber) {
+      if (!treeCache_->IsAsyncReading() && enableTriggerCache_ && (trainedSet_.find(branch) == trainedSet_.end())) {
+        return checkTriggerCacheImpl(branch, entryNumber);
+      } else {
+        return nullptr;
+      }
+    }
+
+    // If this function is called, we already know
+    // the trigger cache is active and it was a cache miss for the regular cache.
+    TTreeCache* SparseReadCache::checkTriggerCacheImpl(TBranch* branch, EntryNumber entryNumber) {
+      // This branch is not going to be in the cache.
+      // Assume this is a "trigger pattern".
+      // Always make sure the branch is added to the trigger set.
+      if (triggerSet_.find(branch) == triggerSet_.end()) {
+        triggerSet_.insert(branch);
+        if (triggerTreeCache_.get()) {
+          triggerTreeCache_->AddBranch(branch, kTRUE);
+        }
+      }
+
+      if (rawTriggerSwitchOverEntry_ < 0) {
+        // The trigger has never fired before.  Take everything not in the
+        // trainedSet and load it from disk
+
+        // Calculate the end of the next cluster; triggers in the next cluster
+        // will use the triggerCache, not the rawTriggerCache.
+        //
+        // Guarantee that rawTriggerSwitchOverEntry_ is positive (non-zero) after completion
+        // of this if-block.
+        TTree::TClusterIterator clusterIter = tree_->GetClusterIterator(entryNumber);
+        while ((rawTriggerSwitchOverEntry_ < entryNumber) || (rawTriggerSwitchOverEntry_ <= 0)) {
+          rawTriggerSwitchOverEntry_ = clusterIter();
+        }
+
+        // ROOT will automatically expand the cache to fit one cluster; hence, we use
+        // 5 MB as the cache size below
+        rawTriggerTreeCache_ = createCacheWithSize(5 * 1024 * 1024);
+        if (rawTriggerTreeCache_) {
+          rawTriggerTreeCache_->SetEnablePrefetching(false);
+          TObjArray* branches = tree_->GetListOfBranches();
+          int branchCount = branches->GetEntriesFast();
+
+          // Train the rawTriggerCache to have everything not in the regular cache.
+          rawTriggerTreeCache_->SetLearnEntries(0);
+          rawTriggerTreeCache_->SetEntryRange(entryNumber, rawTriggerSwitchOverEntry_);
+          for (int i = 0; i < branchCount; i++) {
+            auto tmp_branch = dynamic_cast<TBranch*>(branches->UncheckedAt(i));
+            if (trainedSet_.find(tmp_branch) != trainedSet_.end()) {
+              continue;
+            }
+            rawTriggerTreeCache_->AddBranch(tmp_branch, kTRUE);
+          }
+          rawTriggerTreeCache_->StopLearningPhase();
+        }
+        performedSwitchOver_ = false;
```

Comment:
Just to note that in the earlier code the `if` behavior was different
https://github.com/cms-sw/cmssw/blob/f934747fc3f3b0c983bafb0b6e8a8571c80f5f3a/IOPool/Input/src/RootTree.cc#L347-L363
but given that `rawTriggerTreeCache_` was used unconditionally after the first `if` maybe that was a bug and the structure here looks like plausible intended behavior.